MovingBlocks/Terasology
# Chunk provider refactoring. Part 2.

### Contains

* Refactoring of ChunkProviders and related code.
* Full implemented chunk loading pipeline. I reworked it.
 (Instead Generation off-thread and "Finalization" off-thread  separatly  and many 
 fields in ChunkProvider)
* Chunk loading pipeline with implementation on CompletableFutures.
* JavaDocs


Nitpics about javadocs are welcome (forgotten javadocs too)

##### In next PR: 
 * Try remove LightMerging from headless server. Seems it is not needed 
 * Try merge RemoteChunkProvider and LocalChunkProvider - deduplicate code.
 * More Solid for ChunkProviders and related classes
 * Jomlify

### How to test

LocalChunkProvider:
1. Create/Load game
2. move somewhere fast - check performance, check logs, check that chunks loaded and looks nice.
3. purge world - may be several times - check that you purge all, logs nice, you haven't something strange
4. move far away and die(or use teleport) - check performance, check logs, check chunks looks nice.

RemoteChunkProvider:
1. create server and connect to it.
2. Repeat all checks above.

Play with friends. RelativeSystem must work nice.

### Outstanding before merging
  - [x] Cleanup code - remove unused
  - [x] Seems, I made some memory leak.
  - [x] Done some ToDos
